test/Concurrency/isolated_parameters.swift

```diff
@@ -174,7 +190,8 @@ extension TestActor {
   func isolatedMethod() { }
   // expected-note@-1{{calls to instance method 'isolatedMethod()' from outside of its actor context are implicitly asynchronous}}
 
-  func isolatedToParameter(_ other: isolated TestActor) {
+  // expected-warning@+1 {{instance method with 'isolated' parameter cannot be 'nonisolated'; this is an error in Swift 6}}{{3-15=}}
+  nonisolated func isolatedToParameter(_ other: isolated TestActor) {
```

Comment:
This one is surprising... why not? `nonisolated` == not on the Test actor that I'm declared "in" + `isolated other` == on that other actor == still just on _one_ specific actor, this looks correct to me?

Rather, an actor member method with an isolated parameter but _without_ nonisolated would be an error -- since 
the "two" isolations then clash.

Did I miss something?

Reply:
I'm not sure I agree with this... model wise an async function must always be on _one_ actor. We discussed this a lot during the initial actors pitch. So this gets us:

- free `func` (isolated: 0) -> not isolated
- free `nonisolated func` (isolated: 0)  -> not isolated (doesn't really change anything)
- free `func` (isolated: 0)  + `isolated param` (isolated: add param) -> isolated to param
- actor `func` (isolated: add self)  -> isolated to self 
- actor  (isolated: add self) `func` + `isolated param` (isolated: add param) -> ❌ error, attempt to isolate to self + param
- `actor` (isolated: add self) + `nonisolated func` (isolated: remove self) + `isolated param` (isolated: add param) -> isolated to param
- free `func` + `isolated param` + `isolated param2` -> ❌  error, attempt to isolate to param + param.
- free `func` (isolated: 0) + `@Global` (isolated: add Global) -> ok
- actor `func` (isolated: add self) + `@Global` (isolated: add Global) -> ❌ bad, two isolations
- actor `func` (isolated: add self) + `nonisolated` (isolated: remove self) + `@Global` (isolated: add Global) -> good, isolated to Global
- actor `func` (isolated: add self) + `nonisolated` (isolated: remove self) + `isolated param` (isolated: add param) + `@Global` (isolated: add Global) -> ❌ bad again, while we removed self, we're back to 2 isolations which is bad - can't be on param and global. 

In a sense "global actor" is the same as `isolated GlobalActor` semantically, it is an "add isolation to this" so as long as the function already was at `0` isolations, we can do it. So it can be a free function, or a nonisolated actor function. Either are okey.

Reply:
```
 actor ... { 
 // weird, but not really wrong
 nonisolated func ready(other: isolated Worker) { } 
```

indeed it is weird. one can certainly interpret it in various ways to make it make sense, but i think the interpretation that you're proposing only serves to make the language more confusing. under that interpretation when people spot a `nonisolated`, they still _have_ to scan the parameter list to see if it _really is_ not isolated to anything. That's my interpretation of the keyword, "not isolated to any actor". it's a powerful and simple thing. it shouldn't be muddied with "turn off inference of isolation because I'm going to write it explicitly". We'd need to make its definition even more complex to exclude stuff like this:

```
actor A {
  @MainActor nonisolated func f() {}
}
```

In theory `nonisolated` should be written on every bare `async` or non-async function, and today we totally do allow that without any diagnostic to remove it. but it's also not needed, so we allow folks to reduce the noise by leaving it off and letting the compiler infer that.

Reply:
I can't say I really agree with this read of `nonisolated`... Proposal text wise it's always been about "disable the implicit isolated self":

> Instance declarations on an actor type implicitly have an isolated self. However, one can disable this implicit behavior using the nonisolated keyword:

Though we never covered this weird cases in the proposal, so that's quite an omission in the design there tbh... 

On the other hand though, the use case is definitely very questionable, so not worth pushing harder on 🤔 Let's go with your proposed strict semantics and if it ever shows up in a real use case (vs. just me overworrying it might), then we could re-visit, since opening up a restriction is possible in the future..

Looks good then, thanks for discussing!